File: cmd/kseal-cli/internal/cli/masvs.go
package cli

import (
	"encoding/json"
	"sort"
	"strings"

	ksealv1 "github.com/kennguy3n/kseal/server/gen/kseal/v1"
)

// masvsCategories is the ordered set of OWASP MASVS categories the report maps
// the protected build's modules onto. Order mirrors docs/masvs-mapping.md so
// the report reads like the canonical mapping doc.
var masvsCategories = []string{
	"STORAGE", "CRYPTO", "AUTH", "NETWORK", "PLATFORM", "CODE", "RESILIENCE", "PRIVACY",
}

// moduleMASVS maps a hardening/RASP module (as named in a build manifest) to the
// MASVS categories it contributes evidence toward. Keys are normalized
// (lower-case, non-alphanumeric stripped) so manifest variants like
// "anti-hooking", "anti_hooking", and "antiHooking" all resolve. This mirrors
// the control mapping in docs/masvs-mapping.md; it is intentionally client-side
// data so it can track the doc without a server change.
var moduleMASVS = map[string][]string{
	"integrity":      {"CODE", "RESILIENCE"},
	"appintegrity":   {"CODE", "RESILIENCE"},
	"rasp":           {"PLATFORM", "RESILIENCE"},
	"attestation":    {"AUTH", "NETWORK"},
	"apiattestation": {"AUTH", "NETWORK"},
	"network":        {"NETWORK"},
	"tls":            {"NETWORK"},
	"obfuscation":    {"CODE", "RESILIENCE"},
	"antihooking":    {"RESILIENCE"},
	"hooking":        {"RESILIENCE"},
	"environment":    {"PLATFORM", "RESILIENCE"},
	"root":           {"PLATFORM", "RESILIENCE"},
	"jailbreak":      {"PLATFORM", "RESILIENCE"},
	"storage":        {"STORAGE"},
	"crypto":         {"CRYPTO"},
	"privacy":        {"PRIVACY"},
}

// normalizeModule lower-cases a module name and strips any non-alphanumeric
// characters so manifest naming variants resolve to one mapping key.
func normalizeModule(s string) string {
	var b strings.Builder
	for _, r := range strings.ToLower(s) {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// buildManifestProvenance is the subset of a build manifest the MASVS report
// reads. Both `modules` and `modules_enabled` are accepted (build plugins and
// protection profiles use different keys); `transforms` is surfaced verbatim as
// supporting provenance.
type buildManifestProvenance struct {
	Modules        []string `json:"modules"`
	ModulesEnabled []string `json:"modules_enabled"`
	Transforms     []string `json:"transforms"`
}

// MASVSCategoryCoverage is one MASVS category's coverage in a build report.
type MASVSCategoryCoverage struct {
	Category string   `json:"category"`
	Covered  bool     `json:"covered"`
	Modules  []string `json:"modules"`
}

// MASVSReport is the build's MASVS evidence: the build-hash proof, the manifest
// module/transform provenance, and per-category coverage derived from the
// module set. Gaps and notes make the limits of the evidence explicit (no PII,
// fail-honest about what the registry RPCs do and do not expose).
type MASVSReport struct {
	BuildID         string                  `json:"build_id"`
	AppID           string                  `json:"app_id"`
	BuildHash       string                  `json:"build_hash"`
	VersionName     string                  `json:"version_name"`
	VersionCode     int64                   `json:"version_code"`
	Modules         []string                `json:"modules"`
	Transforms      []string                `json:"transforms"`
	Categories      []MASVSCategoryCoverage `json:"categories"`
	CoveredCount    int                     `json:"covered_count"`
	TotalCategories int                     `json:"total_categories"`
	Gaps            []string                `json:"gaps"`
	Notes           []string                `json:"notes"`
}

// buildMASVSReport derives a MASVS evidence report from a registered build,
// reading only the build-hash proof and manifest provenance the registry RPC
// exposes. A missing or unparseable manifest is not an error: the report still
// renders the build proof and records the absence as a note, so the command is
// always informative.
func buildMASVSReport(b *ksealv1.Build) MASVSReport {
	rep := MASVSReport{
		BuildID:         b.GetId(),
		AppID:           b.GetAppId(),
		BuildHash:       b.GetBuildHash(),
		VersionName:     b.GetVersionName(),
		VersionCode:     b.GetVersionCode(),
		TotalCategories: len(masvsCategories),
	}

	manifest := strings.TrimSpace(b.GetManifest())
	if manifest == "" {
		rep.Notes = append(rep.Notes, "build manifest is empty: no module provenance to map; only the build-hash proof is available")
	} else {
		var prov buildManifestProvenance
		if err := json.Unmarshal([]byte(manifest), &prov); err != nil {
			rep.Notes = append(rep.Notes, "build manifest is not valid JSON: module coverage could not be derived")
		} else {
			rep.Modules = dedupeSorted(append(append([]string(nil), prov.Modules...), prov.ModulesEnabled...))
			rep.Transforms = dedupeSorted(prov.Transforms)
		}
	}

	// Accumulate the contributing modules per MASVS category.
	byCategory := map[string]map[string]struct{}{}
	for _, cat := range masvsCategories {
		byCategory[cat] = map[string]struct{}{}
	}
	for _, m := range rep.Modules {
		cats, ok := moduleMASVS[normalizeModule(m)]
		if !ok {
			rep.Notes = append(rep.Notes, "module \""+m+"\" is not mapped to a MASVS category")
			continue
		}
		for _, cat := range cats {
			byCategory[cat][m] = struct{}{}
		}
	}

	for _, cat := range masvsCategories {
		mods := setToSorted(byCategory[cat])
		covered := len(mods) > 0
		rep.Categories = append(rep.Categories, MASVSCategoryCoverage{Category: cat, Covered: covered, Modules: mods})
		if covered {
			rep.CoveredCount++
		} else {
			rep.Gaps = append(rep.Gaps, cat)
		}
	}

	rep.Notes = append(rep.Notes,
		"evidence is derived from the registered build-manifest module set and the build-hash proof; the registry/query RPCs do not expose per-control MASTG verification status or signed attestation artifacts")
	return rep
}

func dedupeSorted(in []string) []string {
	seen := map[string]struct{}{}
	out := make([]string, 0, len(in))
	for _, s := range in {
		s = strings.TrimSpace(s)
		if s == "" {
			continue
		}
		if _, ok := seen[s]; ok {
			continue
		}
		seen[s] = struct{}{}
		out = append(out, s)
	}
	sort.Strings(out)
	return out
}

func setToSorted(set map[string]struct{}) []string {
	out := make([]string, 0, len(set))
	for s := range set {
		out = append(out, s)
	}
	sort.Strings(out)
	return out
}

File: web/console/src/lib/masvs.ts
import type { Build } from "../gen/kseal/v1/registry_pb";

// MASVS evidence, computed client-side from a registered build's manifest — the
// same derivation the CLI report generator does (cmd/kseal-cli masvs.go) so the
// console and the CLI agree without a server round-trip. The build proof and
// module/transform provenance come straight from RegistryService builds, which
// the console already reads, so this view is real data today (no dependency on
// the WS-K RPCs).

// Ordered MASVS categories, mirroring docs/masvs-mapping.md.
export const MASVS_CATEGORIES = [
  "STORAGE",
  "CRYPTO",
  "AUTH",
  "NETWORK",
  "PLATFORM",
  "CODE",
  "RESILIENCE",
  "PRIVACY",
] as const;

export type MasvsCategory = (typeof MASVS_CATEGORIES)[number];

// Maps a hardening/RASP module to the MASVS categories it evidences. Keys are
// normalized (lower-case, non-alphanumeric stripped) so manifest naming
// variants ("anti-hooking", "anti_hooking", "antiHooking") all resolve. Mirrors
// moduleMASVS in cmd/kseal-cli/internal/cli/masvs.go.
const MODULE_MASVS: Record<string, MasvsCategory[]> = {
  integrity: ["CODE", "RESILIENCE"],
  appintegrity: ["CODE", "RESILIENCE"],
  rasp: ["PLATFORM", "RESILIENCE"],
  attestation: ["AUTH", "NETWORK"],
  apiattestation: ["AUTH", "NETWORK"],
  network: ["NETWORK"],
  tls: ["NETWORK"],
  obfuscation: ["CODE", "RESILIENCE"],
  antihooking: ["RESILIENCE"],
  hooking: ["RESILIENCE"],
  environment: ["PLATFORM", "RESILIENCE"],
  root: ["PLATFORM", "RESILIENCE"],
  jailbreak: ["PLATFORM", "RESILIENCE"],
  storage: ["STORAGE"],
  crypto: ["CRYPTO"],
  privacy: ["PRIVACY"],
};

export interface MasvsCategoryCoverage {
  category: MasvsCategory;
  covered: boolean;
  modules: string[];
}

export interface MasvsReport {
  buildId: string;
  appId: string;
  buildHash: string;
  versionName: string;
  versionCode: number;
  modules: string[];
  transforms: string[];
  categories: MasvsCategoryCoverage[];
  coveredCount: number;
  totalCategories: number;
  gaps: MasvsCategory[];
  notes: string[];
}

interface ManifestProvenance {
  modules?: unknown;
  modules_enabled?: unknown;
  transforms?: unknown;
}

function normalizeModule(s: string): string {
  let out = "";
  for (const ch of s.toLowerCase()) {
    if ((ch >= "a" && ch <= "z") || (ch >= "0" && ch <= "9")) out += ch;
  }
  return out;
}

function dedupeSorted(values: string[]): string[] {
  const seen = new Set<string>();
  for (const raw of values) {
    const s = raw.trim();
    if (s) seen.add(s);
  }
  return [...seen].sort();
}

// Defensive: a manifest field may be absent or the wrong type. Coerce only
// genuine string arrays; anything else contributes nothing.
function asStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((v): v is string => typeof v === "string");
}

// Derives a MASVS evidence report from a registered build. A missing or
// unparseable manifest is not an error: the report still renders the build
// proof and records the absence as a note, so the view is always informative.
export function buildMasvsReport(build: Build): MasvsReport {
  const notes: string[] = [];
  let modules: string[] = [];
  let transforms: string[] = [];

  const manifest = build.manifest.trim();
  if (manifest === "") {
    notes.push(
      "build manifest is empty: no module provenance to map; only the build-hash proof is available",
    );
  } else {
    try {
      const prov = JSON.parse(manifest) as ManifestProvenance;
      modules = dedupeSorted([
        ...asStringArray(prov.modules),
        ...asStringArray(prov.modules_enabled),
      ]);
      transforms = dedupeSorted(asStringArray(prov.transforms));
    } catch {
      notes.push(
        "build manifest is not valid JSON: module coverage could not be derived",
      );
    }
  }

  const byCategory: Record<MasvsCategory, Set<string>> = {
    STORAGE: new Set(),
    CRYPTO: new Set(),
    AUTH: new Set(),
    NETWORK: new Set(),
    PLATFORM: new Set(),
    CODE: new Set(),
    RESILIENCE: new Set(),
    PRIVACY: new Set(),
  };

  for (const m of modules) {
    const cats = MODULE_MASVS[normalizeModule(m)];
    if (!cats) {
      notes.push(`module "${m}" is not mapped to a MASVS category`);
      continue;
    }
    for (const cat of cats) byCategory[cat].add(m);
  }

  const categories: MasvsCategoryCoverage[] = [];
  const gaps: MasvsCategory[] = [];
  let coveredCount = 0;
  for (const cat of MASVS_CATEGORIES) {
    const mods = [...byCategory[cat]].sort();
    const covered = mods.length > 0;
    categories.push({ category: cat, covered, modules: mods });
    if (covered) coveredCount++;
    else gaps.push(cat);
  }

  notes.push(
    "evidence is derived from the registered build-manifest module set and the build-hash proof; the registry/query RPCs do not expose per-control MASTG verification status or signed attestation artifacts",
  );

  return {
    buildId: build.id,
    appId: build.appId,
    buildHash: build.buildHash,
    versionName: build.versionName,
    versionCode: Number(build.versionCode),
    modules,
    transforms,
    categories,
    coveredCount,
    totalCategories: MASVS_CATEGORIES.length,
    gaps,
    notes,
  };
}
